File: tests/python/test_player_inventory.py
import unittest

from endstone_blockdata.player_inventory import (
    LivePlayerInventoryAdapter,
    PlayerInventorySection,
    PlayerInventorySnapshot,
    PlayerInventoryView,
)


def item(identifier: str, count: int = 1):
    return {"Name": identifier, "Count": count}


class FakeBridge:
    def __init__(self, snapshot):
        self.snapshot = snapshot
        self.last_patch = None

    def player_inventory_available(self, _server):
        return True

    def player_inventory_capabilities(self, _server):
        return {
            "adapter": "fake",
            "main": True,
            "armor": True,
            "offhand": True,
            "ender_chest": True,
            "item_user_nbt": True,
        }

    def capture_player_inventory(self, _server, _player):
        return self.snapshot

    def apply_player_inventory(self, _server, _player, patch, policy):
        self.last_patch = patch
        return {
            "ok": True,
            "status": "applied",
            "message": policy,
            "resulting_revision": 124,
        }


class PlayerInventoryTests(unittest.TestCase):
    def setUp(self):
        bundle = item("minecraft:bundle")
        bundle["tag"] = {
            "storage_item_component_content": [
                {**item("minecraft:diamond", 3), "Slot": 0}
            ]
        }
        self.raw = {
            "player_name": "Josh",
            "xuid": "1234",
            "selected_hotbar_slot": 2,
            "main_size": 36,
            "armor_size": 4,
            "offhand_size": 1,
            "ender_chest_size": 27,
            "main": [
                {"slot": 2, "item": bundle, "revision": 11},
                {"slot": 8, "item": item("minecraft:stone", 64), "revision": 12},
            ],
            "armor": [
                {"slot": 0, "item": item("minecraft:diamond_helmet"), "revision": 13}
            ],
            "offhand": [
                {"slot": 0, "item": item("minecraft:shield"), "revision": 14}
            ],
            "ender_chest": [],
            "revision": 123,
        }
        self.snapshot = PlayerInventorySnapshot.from_mapping(self.raw)
        self.view = PlayerInventoryView(self.snapshot)

    def test_reads_all_sections(self):
        self.assertEqual(self.view.get_item(PlayerInventorySection.MAIN, 8)["Count"], 64)
        self.assertEqual(self.view.get_item(PlayerInventorySection.ARMOR, 0)["Name"],
                         "minecraft:diamond_helmet")
        self.assertEqual(self.view.get_item(PlayerInventorySection.OFFHAND, 0)["Name"],
                         "minecraft:shield")
        self.assertIsNone(self.view.get_item(PlayerInventorySection.ENDER_CHEST, 4))

    def test_finds_and_patches_bundle_in_player_inventory(self):
        refs = self.view.find_storage_items()
        self.assertEqual(len(refs), 1)
        self.assertEqual(refs[0].section, PlayerInventorySection.MAIN)
        self.assertEqual(refs[0].slot, 2)

        bundle = self.view.storage_item(PlayerInventorySection.MAIN, 2)
        bundle.set_item(1, item("minecraft:emerald", 4))
        patch = self.view.patch_storage_item(PlayerInventorySection.MAIN, 2, bundle)
        self.assertEqual(patch.expected_revision, 123)
        contents = patch.main_updates[2]["tag"]["storage_item_component_content"]
        self.assertEqual(len(contents), 2)

    def test_armor_offhand_and_ender_chest_patches(self):
        armor = self.view.patch_item(PlayerInventorySection.ARMOR, 1,
                                     item("minecraft:diamond_chestplate"))
        self.assertIn(1, armor.armor_updates)

        offhand = self.view.clear_item(PlayerInventorySection.OFFHAND, 0)
        self.assertEqual(offhand.offhand_removals, {0})

        ender = self.view.patch_item(PlayerInventorySection.ENDER_CHEST, 5,
                                    item("minecraft:gold_ingot", 8))
        self.assertIn(5, ender.ender_chest_updates)

    def test_slot_bounds(self):
        with self.assertRaises(IndexError):
            self.view.get_item(PlayerInventorySection.ARMOR, 4)
        with self.assertRaises(IndexError):
            self.view.patch_item(PlayerInventorySection.OFFHAND, 1, item("minecraft:stone"))

    def test_live_bridge_wrapper(self):
        bridge = FakeBridge(self.raw)
        adapter = LivePlayerInventoryAdapter(object(), bridge)
        self.assertTrue(adapter.available)
        self.assertEqual(adapter.capabilities()["adapter"], "fake")
        captured = adapter.capture(object())
        self.assertEqual(captured.player_name, "Josh")
        patch = PlayerInventoryView(captured).clear_item(PlayerInventorySection.MAIN, 8)
        result = adapter.apply(object(), patch)
        self.assertTrue(result["ok"])
        self.assertEqual(bridge.last_patch["main_removals"], [8])


if __name__ == "__main__":
    unittest.main()

File: tests/cpp/test_player_inventory.cpp
#include "endstone_blockdata/player_inventory.h"
#include "endstone_blockdata/storage_item.h"

#include <algorithm>
#include <cassert>
#include <iostream>
#include <stdexcept>

using namespace endstone_blockdata;

namespace endstone {
class Player {};
}

namespace {

class FakePlayerInventoryAdapter final : public IPlayerInventoryAdapter {
public:
    std::optional<PlayerInventorySnapshot> capture(endstone::Player &) override
    {
        return snapshot;
    }

    ApplyResult apply(
        endstone::Player &,
        const PlayerInventoryPatch &patch,
        ConflictPolicy policy) override
    {
        last_patch = patch;
        last_policy = policy;
        ++apply_calls;
        return {ApplyStatus::Applied, "applied", 999};
    }

    std::string_view name() const noexcept override { return "fake-player-inventory"; }

    PlayerInventorySnapshot snapshot;
    PlayerInventoryPatch last_patch;
    ConflictPolicy last_policy{ConflictPolicy::FailIfChanged};
    int apply_calls{};
};

NbtValue item(std::string name, std::int8_t count = 1)
{
    return NbtValue::compound({
        {"Name", std::move(name)},
        {"Count", count},
    });
}

} // namespace

int main()
{
    PlayerInventorySnapshot snapshot;
    snapshot.player_name = "Josh";
    snapshot.xuid = "1234";
    snapshot.selected_hotbar_slot = 2;
    snapshot.main_size = 36;
    snapshot.armor_size = 4;
    snapshot.offhand_size = 1;
    snapshot.ender_chest_size = 27;

    auto bundle = item("minecraft:bundle");
    StorageItemView bundle_view(bundle, {}, true);
    bundle_view.setSlot(0, item("minecraft:diamond", 3));
    bundle = bundle_view.item();

    snapshot.main.push_back({2, bundle, hashNbt(bundle)});
    snapshot.main.push_back({8, item("minecraft:stone", 64), 0});
    snapshot.main.back().revision = hashNbt(snapshot.main.back().item);
    snapshot.armor.push_back({0, item("minecraft:diamond_helmet"), 0});
    snapshot.armor.back().revision = hashNbt(snapshot.armor.back().item);
    snapshot.offhand.push_back({0, item("minecraft:shield"), 0});
    snapshot.offhand.back().revision = hashNbt(snapshot.offhand.back().item);
    snapshot.revision = calculatePlayerInventoryRevision(snapshot);

    PlayerInventorySnapshot reordered = snapshot;
    std::reverse(reordered.main.begin(), reordered.main.end());
    assert(calculatePlayerInventoryRevision(reordered) == snapshot.revision);

    PlayerInventoryView view(snapshot);
    const auto bundle_slot = view.getItem(PlayerInventorySection::Main, 2);
    assert(bundle_slot.has_value());
    assert(isStorageItemNbt(bundle_slot->item));

    StorageItemView editable(bundle_slot->item);
    editable.setSlot(1, item("minecraft:emerald", 4));
    auto bundle_patch = view.patchItem(PlayerInventorySection::Main, 2, editable.item());
    assert(bundle_patch.expected_revision == snapshot.revision);
    assert(bundle_patch.main_updates.contains(2));

    auto armor_patch = view.patchItem(
        PlayerInventorySection::Armor,
        1,
        item("minecraft:diamond_chestplate"));
    assert(armor_patch.armor_updates.contains(1));

    auto offhand_patch = view.clearItem(PlayerInventorySection::Offhand, 0);
    assert(offhand_patch.offhand_removals.contains(0));

    auto ender_patch = view.patchItem(
        PlayerInventorySection::EnderChest,
        5,
        item("minecraft:gold_ingot", 8));
    assert(ender_patch.ender_chest_updates.contains(5));

    bool threw = false;
    try {
        static_cast<void>(view.patchItem(
            PlayerInventorySection::Armor,
            4,
            item("minecraft:stone")));
    }
    catch (const std::out_of_range &) {
        threw = true;
    }
    assert(threw);

    auto changed = snapshot;
    changed.offhand.clear();
    assert(calculatePlayerInventoryRevision(changed) != snapshot.revision);

    auto fake = std::make_shared<FakePlayerInventoryAdapter>();
    fake->snapshot = snapshot;
    PlayerInventoryService service(fake);
    endstone::Player player;

    PlayerInventoryPatch invalid_offhand;
    invalid_offhand.offhand_removals.insert(1);
    assert(service.apply(player, invalid_offhand).status == ApplyStatus::InvalidPatch);
    assert(fake->apply_calls == 0);

    PlayerInventoryPatch contradictory;
    contradictory.main_updates.emplace(0, PlayerInventoryItemSnapshot{0, item("minecraft:stone"), 0});
    contradictory.main_removals.insert(0);
    assert(service.apply(player, contradictory).status == ApplyStatus::InvalidPatch);
    assert(fake->apply_calls == 0);

    PlayerInventoryPatch valid;
    valid.expected_revision = snapshot.revision;
    valid.main_updates.emplace(1, PlayerInventoryItemSnapshot{1, item("minecraft:diamond"), 0});
    const auto result = service.apply(player, valid, ConflictPolicy::FailIfChanged);
    assert(result.ok());
    assert(fake->apply_calls == 1);
    assert(fake->last_patch.main_updates.contains(1));
    assert(service.adapterName() == "fake-player-inventory");

    assert(service.apply(player, valid, ConflictPolicy::Replace).status == ApplyStatus::Unsupported);
    assert(fake->apply_calls == 1);

    std::cout << "player inventory tests passed\n";
    return 0;
}
